File: web/antcode-frontend/src/services/logs.ts
import { BaseService } from './base'
import apiClient from './api'
import { createLogStreamConnection } from './logStreamConnection'

export type LogLevel = 'DEBUG' | 'INFO' | 'WARNING' | 'ERROR' | 'CRITICAL'
export type LogType = 'stdout' | 'stderr' | 'system' | 'application'
export type LogFormat = 'structured' | 'raw'

export interface LogEntry {
  id?: string
  timestamp: string
  level: LogLevel
  log_type: LogType
  run_id?: string
  task_id?: string
  message: string
  source?: string
  file_path?: string
  line_number?: number
  sequence?: number
  extra_data?: Record<string, unknown>
}

export interface StructuredLogData {
  total: number
  page: number
  size: number
  items: LogEntry[]
}

export interface UnifiedLogResponse {
  run_id: string
  format: LogFormat
  log_type?: string
  raw_content?: string
  file_path?: string
  file_size?: number
  lines_count?: number
  last_modified?: string
  structured_data?: StructuredLogData
}

export interface UnifiedLogParams {
  run_id: string
  format?: LogFormat
  log_type?: 'stdout' | 'stderr'
  level?: LogLevel
  lines?: number
  search?: string
}

export interface LogQueryParams {
  run_id?: string
  log_type?: 'stdout' | 'stderr'
  level?: LogLevel
  lines?: number
  search?: string
  page?: number
  size?: number
  start_time?: string
  end_time?: string
  task_id?: string
}

export interface LogFileResponse {
  success: boolean
  code: number
  message: string
  data: {
    run_id: string
    log_type: string
    content: string
    file_path: string
    file_size: number
    lines_count: number
    last_modified?: string
  }
}

export interface LogListResponse {
  success: boolean
  code: number
  message: string
  data: StructuredLogData
}

export interface LogStreamConnection {
  disconnect: () => void
}

export type HistoricalLogsPhase = 'loading' | 'loaded' | 'empty' | 'gap'

export interface HistoricalLogsUpdate {
  phase: HistoricalLogsPhase
  sentLines?: number
  truncated?: boolean
}

export interface LogStreamOptions {
  runId: string
  onMessage?: (log: LogEntry) => void
  onError?: (error: unknown) => void
  onStateChange?: (state: string) => void
  onStatusUpdate?: (status: { status: string; message?: string; progress?: number }) => void
  onHistoricalLogsUpdate?: (update: HistoricalLogsUpdate) => void
}

class LogService extends BaseService {
  constructor() {
    super('/api/v1')
  }

  async getUnifiedLogs(params: UnifiedLogParams): Promise<UnifiedLogResponse> {
    const queryParams: Record<string, string | number> = {
      format: params.format || 'structured'
    }

    if (params.log_type) queryParams.log_type = params.log_type
    if (params.level) queryParams.level = params.level
    if (params.lines) queryParams.lines = Math.min(Math.max(params.lines, 1), 10000)
    if (params.search) queryParams.search = params.search

    return await this.get<UnifiedLogResponse>(`/logs/runs/${params.run_id}`, { params: queryParams })
  }

  async getRunLogs(runId: string, params?: LogQueryParams): Promise<LogListResponse> {
    const unified = await this.getUnifiedLogs({
      run_id: runId,
      format: 'structured',
      log_type: params?.log_type,
      level: params?.level,
      lines: params?.lines,
      search: params?.search,
    })

    if (!unified.structured_data) {
      throw new Error('日志接口返回缺少 structured_data')
    }

    return {
      success: true,
      code: 200,
      message: '获取成功',
      data: unified.structured_data,
    }
  }

  async getStdoutLogs(runId: string, lines?: number): Promise<LogFileResponse> {
    const params: Record<string, number> = {}
    if (lines) params.lines = Math.min(Math.max(lines, 1), 10000)

    const unified = await this.get<UnifiedLogResponse>(`/logs/runs/${runId}/stdout`, { params })

    return {
      success: true,
      code: 200,
      message: '获取成功',
      data: {
        run_id: unified.run_id,
        log_type: unified.log_type || 'stdout',
        content: unified.raw_content || '',
        file_path: unified.file_path || '',
        file_size: unified.file_size || 0,
        lines_count: unified.lines_count || 0,
        last_modified: unified.last_modified,
      },
    }
  }

  async getStderrLogs(runId: string, lines?: number): Promise<LogFileResponse> {
    const params: Record<string, number> = {}
    if (lines) params.lines = Math.min(Math.max(lines, 1), 10000)

    const unified = await this.get<UnifiedLogResponse>(`/logs/runs/${runId}/stderr`, { params })

    return {
      success: true,
      code: 200,
      message: '获取成功',
      data: {
        run_id: unified.run_id,
        log_type: unified.log_type || 'stderr',
        content: unified.raw_content || '',
        file_path: unified.file_path || '',
        file_size: unified.file_size || 0,
        lines_count: unified.lines_count || 0,
        last_modified: unified.last_modified,
      },
    }
  }

  /**
   * 申请日志流一次性 ticket（避免把长期 JWT 写入 URL）。
   * 原生 EventSource 无法携带 Authorization 头，故用 60s TTL 的一次性票据换取接入；
   * 后端 401 时由 apiClient 响应拦截器统一处理（刷新 token + 重试）。
   */
  async getStreamTicket(runId: string): Promise<string> {
    const response = await apiClient.post<{ data?: { ticket?: string }; ticket?: string }>(
      '/api/v1/logs/stream-ticket',
      undefined,
      { params: { run_id: runId } },
    )
    const body = response.data
    const ticket = body?.data?.ticket ?? body?.ticket
    if (!ticket) {
      throw new Error('日志流 ticket 接口未返回 ticket')
    }
    return ticket
  }

  connectLogStream(options: LogStreamOptions): LogStreamConnection | null {
    if (!options.runId) {
      options.onError?.('runId is required for log stream connection')
      return null
    }
    return createLogStreamConnection(() => this.getStreamTicket(options.runId), options)
  }
}

export const logService = new LogService()
export default logService

File: packages/antcode_core/src/antcode_core/domain/schemas/logs.py
"""
日志相关的Pydantic模式定义
包含日志查询、响应等数据模式
"""

from datetime import datetime
from enum import StrEnum
from typing import Any

from pydantic import BaseModel, Field


class LogLevel(StrEnum):
    """日志级别枚举"""

    DEBUG = "DEBUG"
    INFO = "INFO"
    WARNING = "WARNING"
    ERROR = "ERROR"
    CRITICAL = "CRITICAL"


class LogType(StrEnum):
    """日志类型枚举"""

    STDOUT = "stdout"
    STDERR = "stderr"
    SYSTEM = "system"
    APPLICATION = "application"


class LogFormat(StrEnum):
    """日志输出格式枚举"""

    STRUCTURED = "structured"  # 结构化JSON格式
    RAW = "raw"  # 原始文本格式


class LogEntry(BaseModel):
    """日志条目模型"""

    id: int = 0
    timestamp: datetime
    level: LogLevel
    log_type: LogType
    run_id: str = ""
    task_id: str = Field("", description="任务公开ID")
    message: str
    source: str = ""
    line_number: int = 0
    extra_data: dict[str, Any] = Field(default_factory=dict)


class LogQueryParams(BaseModel):
    """日志查询参数"""

    page: int = Field(1, ge=1, description="页码")
    size: int = Field(50, ge=1, le=1000, description="每页数量")
    level: LogLevel | None = Field(None, description="日志级别过滤")
    log_type: LogType | None = Field(None, description="日志类型过滤")
    run_id: str | None = Field(None, description="运行ID过滤")
    task_id: str | None = Field(None, description="任务公开ID过滤")
    start_time: datetime | None = Field(None, description="开始时间")
    end_time: datetime | None = Field(None, description="结束时间")
    search: str | None = Field(None, description="搜索关键词")


class LogListResponse(BaseModel):
    """日志列表响应"""

    total: int
    page: int
    size: int
    items: list[LogEntry]


class UnifiedLogResponse(BaseModel):
    """统一日志响应（支持多种格式）"""

    run_id: str
    format: LogFormat
    log_type: str = ""

    # 结构化格式字段
    structured_data: LogListResponse = Field(
        default_factory=lambda: LogListResponse(total=0, page=1, size=50, items=[])
    )

    # 原始格式字段
    raw_content: str = ""
    file_size: int = 0
    lines_count: int = 0
